### ISSUE TITLE
Lambda invokeAsync not working with params

### ISSUE THREAD

**Issue by ramonck** [aws/aws-sdk-js#478]
Srs.,

I'm trying to invoke a function with params and I'm getting the 202 "Default" response which is ok, and when I look at CloudWatch it tells me it was not able to fetch the first parameter as an error of execution with lambdaClient, please tell me what's wrong.

var params = {
        FunctionName: functionName,
        InvokeArgs: JSON.stringify({to: ['*******'], from:'**********', subject:'******', text_message:'*****'})
      }
      console.log(invokeArgs);
      this.lambdaClient.invokeAsync(params, function(err, data) {
            if (err) { throw 'Erro: '+err } else { console.log(data);}
      });

Error message:
Failure while running task: TypeError: Cannot read property 'to' of undefined at /var/task/index.js:22:17 at exports.handler (/var/task/index.js:68:3)

**Comment by lsegal** [comment]
It looks like the error is coming from your function handler, not the API call. Can you list portions of the handler function itself? It looks like the structure you are calling `.to` on is undefined.

**Comment by ramonck** [comment]
Hi,

The function works fine within NodeJS, but it's not working from web interface.
# 
    var AWS = require("aws-sdk");
    exports.handler = function(event, context) {
      var _callbacked = false;
      var _callback = function(retval) {
    if (_callbacked) { return; }
    _callbacked = true;
    var _sqsOptions = {"accessKeyId":"[ACCESSKEY]","secretAccessKey":"[SECRETKEY]","region":"us-east-1","Role":"arn:aws:iam::9999999:role/lambda_exec_role","FunctionName":"MYFUNCTIONNAME","Description":"send Email","QueueUrl":"https://sqs.us-east-1.amazonaws.com/99999999/LambdaResultsQueue"};
    var _sqsClient = new AWS.SQS(_sqsOptions);
    var _sqsParams = {
      QueueUrl: "https://sqs.us-east-1.amazonaws.com/9999999999/LambdaResultsQueue",
      MessageBody: JSON.stringify({FunctionName: "MYFUNCTIONAME", ReturnValue: retval, CallId: event.callId})
    }
    _sqsClient.sendMessage(_sqsParams, function(err, data) {
      if (err) { throw err; }
      context.done();
    });
      };
      (function (event, callback) {
      // Running on AWS Lambda
      var aws         = require('aws-sdk');
      var ses = new aws.SES({apiVersion: '2010-12-01'});
      var to = event.to;
      var cc = event.cc;

My desire is to use the official library :8ball:

**Comment by lsegal** [comment]
@ramonck I'm not sure I follow the code sample provided. Is the `(function(event, callback) {` portion related? Only the code inside of the `exports.handler()` will execute when you invoke a lambda function.

Also as a sidenote, you should never hardcode access / secret access keys in your Lambda functions. Instead, rely on the IAM roles that you can configure for the instance. If you already have a role configured, simply ensure that the role has permissions to access your SQS queue and you should be able to omit any credential configuration. The SDK will pickup credentials automatically when using IAM roles.

**Comment by ramonck** [comment]
Hi,

I don't see that as being related to what I'm actually reporting as the problem, I'm having correct permissions to access my function in Lambda from Web and Backend, I understand it as being a better practice your point of view, the problem is exactly in this line ``` var to = event.to;``` in which I'm defining a variable coming in from my JSON invoke parameter and it's not there from the WEB API aws-sdk-js and in the backend api works fine with the same parameters, that's the problem.

**Comment by lsegal** [comment]
> I don't see that as being related to what I'm actually reporting as the problem, 

Correct, it's simply a sidenote.

Is the code you pasted exactly the function you have logged in Lambda? If so, it looks like there is a syntax error on my end, or, at the very least, a semantic error, as the code after the `Running on AWS Lambda` comment is not part of the handler that executes per request. As mentioned, all code in your function has to be inside the single `exports.handler` function to execute per request. 

If you could provide the exact (and complete) code in your function, it would help to diagnose the issue. That said, if the error is that `event` is undefined, that is unlikely an issue with the SDK and more likely one with the way Lambda is processing the `event` parameter on their side.

**Comment by ramonck** [comment]
Hi,

I would understand that's AWS Lambda's parsing problem if the problem also ocurred from the backend, which is not case, it's only happening from the aws-sdk-js, that's why I'm directing the problem through here.

Ok, here's the full code.

```
var AWS = require("aws-sdk");
exports.handler = function(event, context) {
  var _callbacked = false;
  var _callback = function(retval) {
    if (_callbacked) { return; }
    _callbacked = true;
    var _sqsOptions = {"accessKeyId":"ACCESSKEY","secretAccessKey":"SECRETKEY","region":"us-east-1","Role":"arn:aws:iam::9999999999:role/lambda_exec_role","FunctionName":"FUNCTIONAME","Description":"send Email","QueueUrl":"https://sqs.us-east-1.amazonaws.com/999999999/LambdaResultsQueue"};
    var _sqsClient = new AWS.SQS(_sqsOptions);
    var _sqsParams = {
      QueueUrl: "https://sqs.us-east-1.amazonaws.com/99999999/LambdaResultsQueue",
      MessageBody: JSON.stringify({FunctionName: "FUNCTIONNAME", ReturnValue: retval, CallId: event.callId})
    }
    _sqsClient.sendMessage(_sqsParams, function(err, data) {
      if (err) { throw err; }
      context.done();
    });
  };
  (function (event, callback) {
  // Running on AWS Lambda
  var aws         = require('aws-sdk');
  var ses = new aws.SES({apiVersion: '2010-12-01'});
  var to = event.to;
  var cc = event.cc;
  var bcc = event.bcc;
  var from = event.from;
  var subject = event.subject;
  var textMessage = event.text_message;
  var htmlMessage = event.html_message;
  var charSet = "UTF-8";
  var params = {
    Destination: {
      BccAddresses: bcc ? bcc : [],
      CcAddresses: cc ? cc : [],
      ToAddresses: to
    },
    Message: {
      Body: {},
      Subject: {
        Data: subject ? subject : "No subject",
        Charset: charSet
      }
    },
    Source: from,
    ReplyToAddresses: [from]
  };

  if (textMessage != null && textMessage.length > 0) {
    params["Message"]["Body"]["Text"] = {
      Data: textMessage,
      Charset: charSet
    }
  }
  if (htmlMessage != null && htmlMessage.length > 0) {
    params["Message"]["Body"]["Html"] = {
      Data: htmlMessage,
      Charset: charSet
    }
  }

  ses.sendEmail(params, function(err, data) {
    if (err) {
      console.log(err, err.stack); // an error occurred
    } else {
      console.log(data);           // successful response
      context.done();
    }
  });
})(event.args, _callback);
};
```

The invoking part from aws-sdk-js
```
var params = {
        FunctionName: functionName,
        InvokeArgs: JSON.stringify({to: ['NAME@NAME'], from:'NAME@NAME', subject:'MY SUBJECT', text_message:'BODY TEXT MESSAGE'})
      }
      console.log(invokeArgs);
      this.lambdaClient.invokeAsync(params, function(err, data) {
            if (err) { throw 'Erro: '+err } else { console.log(data);}
      });
```

**Comment by lsegal** [comment]
I was able to reproduce your issue from the web console when I used the following JSON payload:

```js
{
  "to": "email@example.org"
}
```

When I inspected `event` it was the exact object as above. I got it to work by providing the following JSON instead:

```js
{
  "args": {
    "to": "email@example.org"
  }
}
```

The payload you craft in the web UI seems to be the exact object that `event` will get-- if your closure is passing in `event.args`, the payload you craft in the web UI has to have `args` as a member of the JSON payload.

Does that help at all?
